server.go

```diff
@@ -1103,20 +1103,38 @@ func chainUnaryServerInterceptors(s *Server) {
 }
 
 func chainUnaryInterceptors(interceptors []UnaryServerInterceptor) UnaryServerInterceptor {
+	// the struct ensures the variables are allocated together, rather than separately, since we
+	// know they should be garbage collected together. This, together with the sync.Pool, saves
+	// 2 allocation and decreases time/call by about 25% on the microbenchmark.
+	type state struct {
+		i            int
+		next         UnaryHandler
+		info         *UnaryServerInfo
+		handler      UnaryHandler
+		interceptors []UnaryServerInterceptor
+	}
+	var statePool = sync.Pool{
+		New: func() interface{} {
+			return &state{}
+		},
+	}
+
 	return func(ctx context.Context, req interface{}, info *UnaryServerInfo, handler UnaryHandler) (interface{}, error) {
-		// the struct ensures the variables are allocated together, rather than separately, since we
-		// know they should be garbage collected together. This saves 1 allocation and decreases
-		// time/call by about 10% on the microbenchmark.
-		var state struct {
-			i    int
-			next UnaryHandler
-		}
+		state := statePool.Get().(*state)
+		defer statePool.Put(state)
```

Comment:
Pardon me for the peanut gallery comment if it is unwelcome (I don't mean it to be, and you are free to ignore it):

Unless I am mistaking, wouldn't it be prudent to reset the state — no pun intended — of the `state` before it gets put back into the pool for two reasons:

1. `state` may hold a live reference to something that would otherwise be eligible for garbage collection — cf. [this note](https://github.com/golang/go/wiki/SliceTricks#delete-without-preserving-order).  And inadvertently we keep more memory alive than strictly necessary. 
2. the array that backs slice for `interceptors []UnaryServerInterceptor` may be larger than necessary and result in extraneous memory pressure if there are multi-modal interceptor counts.  I don't think this is likely a problem here, but a colleague who works on the compiler told me that this is a standing problem with folks using `sync.Pool` with buffer-like types.  I only call that out because the fields in `state` are effectively dynamically sized.

This may apply below.